src/audio/kpb.c

```diff
@@ -356,57 +375,54 @@ static int kpb_copy(struct comp_dev *dev)
  */
 static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source)
 {
-	trace_kpb("kpb_buffer_data()");
 	int size_to_copy = kpb->source_period_bytes;
 	int space_avail;
-	struct history_buffer *hb;
+	struct hb *buff = &kpb->history_buffer;
+	struct hb *first_buff = buff;
 
-#if KPB_NO_OF_HISTORY_BUFFERS == 2
+	trace_kpb("kpb_buffer_data()");
+
+	/* find free buffer */
+	while (buff->next && buff->next != first_buff) {
+		if (buff->state == KPB_BUFFER_FREE)
+			break;
+		else
+			buff = buff->next;
+	}
 
 	while (size_to_copy) {
-		hb = (kpb->his_buf_lp.state == KPB_BUFFER_FREE) ?
-		&kpb->his_buf_lp : &kpb->his_buf_hp;
-		space_avail = (int)hb->end_addr - (int)hb->w_ptr;
+		space_avail = (int)buff->end_addr - (int)buff->w_ptr;
 
 		if (size_to_copy > space_avail) {
```

Comment:
@mrajwa can we add a comment here saying that we're checking for buffer wrap here before copying

Reply:
@mrajwa yep, in general when I see large code blocks like this I cant review them without comments describing what they are doing and how. Always best to add comments within the code, even if it obvious to you as it's not always obvious to others.